library/std/src/sync/reentrant_lock.rs

```diff
@@ -303,18 +385,23 @@ impl<T: ?Sized> Drop for ReentrantLockGuard<'_, T> {
         unsafe {
             *self.lock.lock_count.get() -= 1;
             if *self.lock.lock_count.get() == 0 {
-                self.lock.owner.store(0, Relaxed);
+                self.lock.owner.set(0);
                 self.lock.mutex.unlock();
             }
         }
     }
 }
 
-/// Get an address that is unique per running thread.
+/// Returns the current thread's ThreadId value, which is guaranteed
+/// to be unique across the lifetime of the process.
 ///
-/// This can be used as a non-null usize-sized ID.
-pub(crate) fn current_thread_unique_ptr() -> usize {
-    // Use a non-drop type to make sure it's still available during thread destruction.
-    thread_local! { static X: u8 = const { 0 } }
-    X.with(|x| <*const _>::addr(x))
+/// Panics if called during a TLS destructor on a thread that hasn't
+/// been assigned an ID.
+pub(crate) fn current_thread_id() -> u64 {
+    #[cold]
+    fn no_tid() -> ! {
+        panic!("Thread hasn't been assigned an ID!")
+    }
+
+    crate::thread::try_current_id().map_or_else(|| no_tid(), |tid| tid.as_u64().get())
```

Comment:
I think this is more readable as
```suggestion
    crate::thread::try_current_id().unwrap_or_else(|| no_tid()).as_u64().get()
```